### PR TITLE
Normalize Java SDK (JDK) location logic across the tool

### LINKED ISSUE #122609: JDK actually used for Gradle doesn't appear in `flutter doctor -v` nor in `flutter run -v`
Update: see comments starting at https://github.com/flutter/flutter/issues/122609#issuecomment-1474273992 . Original report follows.

---

After Android configuration changes (creating a new project, adding plugins) I get this error. I am using the latest Android Studio Preview and stable gradle versions. The only way I've found to fix this is to open the Android native part of the project in Android Studio and click the gradle sync button. Either the root issue should be resolved or the Flutter tool should be doing whatever the gradle sync button does automatically.

Related to https://github.com/flutter/flutter/issues/122376

When I try to run a fresh project on an Android device:
```console
Running "flutter pub get" in gradle_test...
Resolving dependencies...
  async 2.10.0 (2.11.0 available)
  characters 1.2.1 (1.3.0 available)
  collection 1.17.0 (1.17.1 available)
  js 0.6.5 (0.6.7 available)
  matcher 0.12.13 (0.12.14 available)
  material_color_utilities 0.2.0 (0.3.0 available)
  meta 1.8.0 (1.9.0 available)
  path 1.8.2 (1.8.3 available)
  test_api 0.4.16 (0.4.18 available)
Got dependencies!
Launching lib/main.dart on sdk gphone64 arm64 in debug mode...
main.dart:1

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/rexios/repos/gradle_test/android/build.gradle' line: 26

* What went wrong:
A problem occurred evaluating root project 'android'.
> A problem occurred configuring project ':app'.
   > Could not open proj generic class cache for build file '/Users/rexios/repos/gradle_test/android/app/build.gradle' (/Users/rexios/.gradle/caches/7.5/scripts/up9lbpef9oco35z135w5mojx).
      > BUG! exception in phase 'semantic analysis' in source unit '_BuildScript_' Unsupported class file major version 63

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 12s
Exception: Gradle task assembleDebug failed with exit code 1
Exited
```

### PR DESCRIPTION
**Pull request by andrewkolos**
Fixes #122609.

Related: https://github.com/flutter/flutter/issues/123643.
Tangentially related: https://github.com/flutter/flutter/issues/122376. (there are probably more issues than this--I need to look around)

CC: @christopherfujino @reidbaker 

## Context 
`flutter doctor -v` prints the path of the `java` binary that is presumably used for all Android-related features (building apps, managing emulators, etc.). However, this logic is different from that used in other places tool outside of `flutter doctor`, which makes it possible for the JDK path that `flutter doctor` lists to be different from the one that flutter actually uses when building Android apps. See [this comment on #122609 from Greg](https://github.com/flutter/flutter/issues/122609#issuecomment-1474293403) and [this comment from me on #122609](https://github.com/flutter/flutter/issues/122609#issuecomment-1496490325) for more info.

This has the potential to be confusing to those debugging build issues in android-enabled flutter projects.

## Change
This change normalizes the logic of JDK-location across the tool. Said another way, any feature that needs to find a JDK installation will use the same logic to do so. The code that implements this logic is also lightly refactored.

## To-do
* Consider restoring usage of `/usr/libexec/java_home` on macOS systems, which was consulted as a second-to-last resort when finding the JDK to use for everything *except* invoking gradle.
* Add tests
* Make the change description above more helpful

## Pre-launch Checklist
<details>

<summary>Pre-launch Checklist</summary>

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
</details>

### CODE REVIEW

File: packages/flutter_tools/lib/src/android/java.dart [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+import '../base/file_system.dart';
+import '../base/os.dart';
+import '../base/platform.dart';
+import 'android_studio.dart';
+
+const String _javaHomeEnvironmentVariable = 'JAVA_HOME';
+const String _javaExecutable = 'java';
+
+/// Attempts to find the home directory of the local Java installation used
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/android/java.dart:13]: nit here and for `findJavaBinary` separate the first sentence of a dartdoc into its own paragraph: https://dart.dev/guides/language/effective-dart/documentation#do-separate-the-first-sentence-of-a-doc-comment-into-its-own-paragraph

File: packages/flutter_tools/lib/src/globals.dart [hunk 2]
```diff
@@ -285,3 +286,12 @@ const String kDefaultFrameworkChannel = 'master';
 
 // Used to build RegExp instances which can detect the VM service message.
 final RegExp kVMServiceMessageRegExp = RegExp(r'The Dart VM service is listening on ((http|//)[a-zA-Z0-9:/=_\-\.\[\]]+)');
+
+String? _javaPath;
+
+String? get javaHome => _javaPath ??= findJavaHome(
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/globals.dart:292]: This is going to lead to a lot more implicit context accesses, does this need to be global?

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/globals.dart:292]: Or, can't we just have whatever needs to access this directly invoke `findJavaHome`?

**Comment by christopherfujino** [comment]
Added a comment to the tracking issue: https://github.com/flutter/flutter/issues/122609#issuecomment-1496841634

**Comment by andrewkolos** [comment]
I am going to mark this as ready-for-review since @christopherfujino has added reviewers--though tests are still needed.

Based on continued discussion on #122609, a simpler change that could be made instead of this one is simply to remove the macOS-specific logic in `findJavaBinary`. This change alone should solve this issue since 1) we already try to use the Android Studio-bundled JDK when trying to run anything dependent on Java[^1] and 2) Android SDK tools already try to use `JAVA_HOME` and `which java`. The only downside there is that the JDK-location logic is not visible at a flutter tool level--it's implicit behavior coming from the Android SDK.

[^1]: see [references to `javaPath` from `android_studio.dart`](https://github.com/flutter/flutter/blob/0cfcdeec89765f5e350137502e3b398538b0a266/packages/flutter_tools/lib/src/android/android_studio.dart#L30)

**Comment by andrewkolos** [comment]
> Based on continued discussion on #122609, a simpler change that could be made instead of this one is simply to remove the macOS-specific logic in `findJavaBinary`. This change alone should solve this issue since 1) we already try to use the Android Studio-bundled JDK when trying to run anything dependent on Java and 2) Android SDK tools already try to use `JAVA_HOME` and `which java`. The only downside there is that the JDK-location logic is not visible at a flutter tool level--it's implicit behavior coming from the Android SDK.

Per this and offline discussion with @christopherfujino. I will make a simpler PR that just removes the macOS-specific logic. This will make code review simpler and cherrypicking easier, which is ideal since we want this change sooner rather than later. Will post another comment with the new PR.

**Comment by andrewkolos** [comment]
Opened https://github.com/flutter/flutter/pull/124233

**Comment by christopherfujino** [comment]
> > Based on continued discussion on #122609, a simpler change that could be made instead of this one is simply to remove the macOS-specific logic in `findJavaBinary`. This change alone should solve this issue since 1) we already try to use the Android Studio-bundled JDK when trying to run anything dependent on Java and 2) Android SDK tools already try to use `JAVA_HOME` and `which java`. The only downside there is that the JDK-location logic is not visible at a flutter tool level--it's implicit behavior coming from the Android SDK.
> 
> Per this and offline discussion with @christopherfujino, I will make a simpler PR that just removes the macOS-specific logic. This will make code review simpler and cherrypicking easier, which is ideal since we want this change sooner rather than later. Will post another comment with the new PR.

Thanks! I still think it's worth landing this refactor, but we can do that once the minimal "fix" has landed.

**Comment by andrewkolos** [comment]
I agree. I will set a reminder to open a tech-debt issue since JDK-finding is still fundamentally confusing at a flutter tool code level

**Comment by christopherfujino** [comment]
> since JDK-finding is still fundamentally confusing at a flutter tool code level

That's an understatement :)

**Pull request closed without merging**